### PR TITLE
scheduler perf: benchstat support

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Makes it possible to compare the performance impact of a code change without relying on the performance dashboard.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

Example output after
```
go test -timeout=0 -count=3 ./test/integration/scheduler_perf/ -short=true -run=none -benchtime=1ns -bench=BenchmarkPerfScheduling/SchedulingBasic >/tmp/log 2>&1
benchstat /tmp/log
```

```
name                                         SchedulingThroughput/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 500 ± 0%

name                                         SchedulingThroughput/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 375 ± 3%

name                                         SchedulingThroughput/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 625 ± 2%

name                                         SchedulingThroughput/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 625 ± 2%

name                                         SchedulingThroughput/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                  983 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 625 ± 2%

name                                         scheduler_framework_extension_point_duration_ms/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.58 ± 2%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.00 ± 3%

name                                         scheduler_framework_extension_point_duration_ms/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.56 ± 1%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.04 ± 2%

name                                         scheduler_framework_extension_point_duration_ms/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.77 ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.51 ± 0%

name                                         scheduler_framework_extension_point_duration_ms/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                 0.81 ± 2%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                1.57 ± 0%

name                                         scheduler_framework_extension_point_duration_ms/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                 2.19 ± 5%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                2.97 ± 9%

name                                         scheduler_pod_scheduling_duration_ms/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                  375 ± 4%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 700 ± 2%

name                                         scheduler_pod_scheduling_duration_ms/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                  348 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                 687 ± 2%

name                                         scheduler_pod_scheduling_duration_ms/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                  956 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                               1.61k ± 5%

name                                         scheduler_pod_scheduling_duration_ms/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                1.12k ± 1%
PerfScheduling/SchedulingBasic/5000Nodes-36                                               2.09k ± 2%

name                                         scheduler_pod_scheduling_duration_ms/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                1.25k ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                               2.47k ± 0%

name                                         scheduler_scheduling_attempt_duration_ms/Average
PerfScheduling/SchedulingBasic/500Nodes-36                                                 3.48 ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                3.97 ± 5%

name                                         scheduler_scheduling_attempt_duration_ms/Perc50
PerfScheduling/SchedulingBasic/500Nodes-36                                                 3.08 ± 0%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                3.76 ± 4%

name                                         scheduler_scheduling_attempt_duration_ms/Perc90
PerfScheduling/SchedulingBasic/500Nodes-36                                                 5.35 ± 3%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                7.30 ± 2%

name                                         scheduler_scheduling_attempt_duration_ms/Perc95
PerfScheduling/SchedulingBasic/500Nodes-36                                                 7.26 ± 1%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                7.80 ± 1%

name                                         scheduler_scheduling_attempt_duration_ms/Perc99
PerfScheduling/SchedulingBasic/500Nodes-36                                                 14.6 ± 4%
PerfScheduling/SchedulingBasic/5000Nodes-36                                                16.2 ±27%
```

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/cc @kerthcet

### CODE REVIEW

**Comment by kerthcet** [comment]
Can this one be independent, or you want to wait until https://github.com/kubernetes/kubernetes/pull/115422 merged.

**Comment by pohly** [comment]
The other PR is needed, otherwise the output is too noisy for benchstat.

**Comment by pohly** [comment]
@kerthcet: I changed this PR so that it contains everything that is needed to generated output that "benchstat" can digest. We could merge it now.

Can you take another look?

**Review by kerthcet — COMMENTED** [review]
Some comments, I'll experiment this tomorrow.

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 1]
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
...
 
+	output := initTestOutput(b)
+
+	// Because we run sequentially, it is possible to change the global
+	// klog logger and redirect log output. Quite a lot of code still uses
+	// it instead of supporting contextual logging.
+	//
+	// Because we leak one goroutine which calls klog, we cannot restore
+	// the previous state.
+	_ = framework.RedirectKlog(b, output)
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:628]: "go test -race" reports a race when changing klog again while the leaked goroutine is running.

File: test/integration/scheduler_perf/main_test.go [hunk 2]
```diff
@@ -18,17 +18,55 @@ package benchmark
...
 	// Run with -v=2, this is the default log level in production.
-	ktesting.DefaultConfig = ktesting.NewConfig(ktesting.Verbosity(2))
-	ktesting.DefaultConfig.AddFlags(flag.CommandLine)
+	ktesting.SetDefaultVerbosity(2)
+
+	// test/integration/framework/flags.go unconditionally initializes the
+	// logging flags. That's correct for most tests, but in the
+	// scheduler_perf test we want more control over the flags, therefore
+	// here strip them out.
+	var fs flag.FlagSet
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/main_test.go:42]: This is a permanent special case for scheduler_perf. As far as I know, it is the only performance benchmark which uses the integration framework.

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 3]
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
 		b.Fatal(err)
 	}
 
+	output := initTestOutput(b)
+
+	// Because we run sequentially, it is possible to change the global
+	// klog logger and redirect log output. Quite a lot of code still uses
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:623]: Hmm, this doesn't sound right to me. I don't think "source code" has a plural form - but I am not a native speaker.

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 4]
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
...
-					dataItems.DataItems = append(dataItems.DataItems, runWorkload(ctx, b, tc, w)...)
+					results := runWorkload(ctx, b, tc, w)
+					dataItems.DataItems = append(dataItems.DataItems, results...)
+
+					if len(results) > 0 {
+						// The default ns/op is not
+						// useful because it includes
+						// the time spent on
+						// initialization and shutdown. Here we suppress it.
+						b.ReportMetric(0, "ns/op")
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:672]: And call that when?

Besides that, the "ns/op" metric is also not very useful. It highly depends on the configuration of the test case.

The new "average pods/s" is a better replacement for a measurement that covers the relevant duration of the measurement section of a test case.

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 5]
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
...
+						// Instead, report the same
+						// results that also get stored
+						// in the JSON file.
+						for _, result := range results {
+							// For some metrics like
+							// scheduler_framework_extension_point_duration_seconds
+							// the actual value has some
+							// other unit. We patch the key
+							// to make it look right.
+							metric := strings.ReplaceAll(result.Labels["Metric"], "_seconds", "_"+result.Unit)
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:683]: This reads from `result`, but doesn't modify it?

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:623]: https://english.stackexchange.com/questions/584324/is-the-phrase-source-code-intrinsically-plural

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 6]
```diff
@@ -579,19 +617,76 @@ func BenchmarkPerfScheduling(b *testing.B) {
...
+					// Ensure that there are no leaked
+					// goroutines.  They could influence
+					// performance of the next benchmark.
+					// This must *after* RedirectKlog
+					// because then during cleanup, the
+					// test will wait for goroutines to
+					// quit *before* restoring klog settings.
+					framework.GoleakCheck(b)
+
+					ctx := context.Background()
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:644]: This context is the parent of the timeout context. Depending on the parameters, it directly replaces the current `context.TODO()` (not using ktesting) or is a context which has the per-test logger (using ktesting).

File: test/integration/scheduler_perf/scheduler_perf_test.go [hunk 7]
```diff
@@ -570,6 +575,39 @@ func (so sleepOp) patchParams(_ *workload) (realOp, error) {
 	return &so, nil
 }
 
+var useTestingLog = flag.Bool("use-testing-log", false, "Write log entries with testing.TB.Log. This is more suitable for unit testing, but less realistic in real benchmarks.")
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/integration/scheduler_perf/scheduler_perf_test.go:578]: Added.

File: test/utils/format/format.go [hunk 8]
```diff
@@ -78,3 +78,8 @@ func useYAML(t reflect.Type) bool {
 		return false
 	}
 }
+
+// AnyToString is a wrapper around format.Object with zero indention.
```
[the remark this answers is not in the document]
Comment (pohly) [reply test/utils/format/format.go]: Removed.

**Comment by kerthcet** [comment]
Besides, do you tidy up the output? Not the same as mine.

**Comment by pohly** [comment]
You need to run `benchstat` on the output of `go test`. I've updated the PR description to show that.

File: test/integration/scheduler_perf/scheduler_perf_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (kerthcet) [reply test/integration/scheduler_perf/scheduler_perf_test.go:683]: I'll take a care of this. 
This PR makes benchmark results visible via benchstat, so
/lgtm

**Pull request merged**